orthanc/orthanc-anon/plugin/pixl.py

```diff
@@ -219,42 +216,151 @@ def OnHeartBeat(output, uri, **request) -> Any:  # noqa: ARG001
+        study_uid,
+        study_resource_id,
+    )
+    orthanc.RestApiDelete(f"/studies/{study_resource_id}")
+
+    # Anonymise the study, re-upload to Orthanc Anon, and notify the export API to export the study
+    with ZipFile(zipped_study_bytes) as zipped_study:
+        anonymised_instances_bytes, anonymised_study_uid = _anonymise_study_instances(
+            zipped_study=zipped_study,
+            study_uid=study_uid,
+        )
+    _upload_instances(anonymised_instances_bytes)
+    anonymised_study_resource_id = _get_existing_study(study_uid=anonymised_study_uid)
+    Send(study_id=anonymised_study_resource_id)
+
+
+def _get_existing_study(study_uid: str) -> Optional[str]:
+    """
+    Get the resource ID for an existing study based on its StudyInstanceUID.
+
+    Returns None if there are no resources with the given StudyInstanceUID.
+    Returns the resource ID if there is a single resource with the given StudyInstanceUID.
+    Returns None if there are multiple resources with the given StudyInstanceUID and deletes
+    the studies.
+    """
+    data = json.dumps(
+        {
+            "Level": "Study",
+            "Query": {
+                "StudyInstanceUID": study_uid,
+            },
+        }
+    )
+    study_resource_ids = json.loads(orthanc.RestApiPost("/tools/find", data))
+    if not study_resource_ids:
+        logger.info(f"No study found with StudyInstanceUID {study_uid}")
+        return None
+
+    if len(study_resource_ids) == 1:
+        return study_resource_ids[0]
+
+    logger.info(
+        f"{len(study_resource_ids)} found with StudyInstanceUID {study_uid}. Deleting studies."
+    )
+
+    return None
+
+
+def _anonymise_study_instances(zipped_study: ZipFile, study_uid: str) -> tuple[list[bytes], str]:
+    """
+    Iterate over all instances and anonymise them.
+
+    Return a list of the bytes of anonymised instances, and the anonymised StudyInstanceUID.
+    """
+    anonymised_instances_bytes = []
+    for file_info in zipped_study.infolist():
+        with zipped_study.open(file_info) as file:
+            logger.debug("Reading file {}", file)
+            try:
+                dataset = dcmread(file)
+            except pydicom.errors.InvalidDicomError:
+                logger.warning("Failed to read file {}.", file)
+                raise
+
+            try:
+                logger.debug("Anonymising file: {}", file)
+                anonymised_instances_bytes.append(_anonymise_dicom_instance(dataset))
+            except PixlDiscardError:
+                pass
+            except Exception:
+                logger.warning("Failed to anonymize file: {}", file)
+                raise
+            else:
+                anonymised_study_uid = dataset[0x0020, 0x000D].value
+
+    if not anonymised_instances_bytes:
+        message = "All instances have been discarded for study {}", study_uid
+        raise ValueError(message)
```

Comment:
I added this check but I'm not sure it's possible for a study to have no instances (unless all the instances are skipped)?

Reply:
yeah all instances could be skipped so seems sensible.